### PR TITLE
Prevent players accessing inventories of other players

### PR DESCRIPTION
**Pull request by appgurueu**
Title says it all. Fixes this cheat. Unfortunately also requires invhack mods to implement workarounds, but they seem to already be doing this.

### CODE REVIEW

**Comment by EliasFleckenstein03** [comment]
Nice, that's another cheat from my hackclient patched (see #10340 )

**Comment by SmallJoker** [comment]
Player inventories are only sent to the correct player. I see this is needed, but not how you'd get that inventory list in first place.

**Comment by rubenwardy** [comment]
> Player inventories are only sent to the correct player. I see this is needed, but not how you'd get that inventory list in first place.

You don't need to have that list, you can guess where items are

**Review by rubenwardy — APPROVED** [review]

File: src/network/serverpackethandler.cpp [hunk 1]
```diff
@@ -629,13 +629,19 @@ void Server::handleCommand_InventoryAction(NetworkPacket* pkt)
...
 			m_inventory_mgr->setInventoryModified(ma->to_inv);
 
-		bool from_inv_is_current_player =
-			(ma->from_inv.type == InventoryLocation::PLAYER) &&
-			(ma->from_inv.name == player->getName());
-
-		bool to_inv_is_current_player =
-			(ma->to_inv.type == InventoryLocation::PLAYER) &&
-			(ma->to_inv.name == player->getName());
+		bool from_inv_is_current_player;
```
Comment (sfan5) [inline src/network/serverpackethandler.cpp:632]: * you need to initialize this to false
* these should also handle `InventoryLocation::CURRENT_PLAYER` (is that ever sent by the client?)

Comment (rubenwardy) [reply src/network/serverpackethandler.cpp:632]: I rely way too much on IDEs for being warned about this

Comment (appgurueu) [reply src/network/serverpackethandler.cpp]: Done.

File: src/network/serverpackethandler.cpp [hunk 2]
```diff
@@ -629,13 +629,21 @@ void Server::handleCommand_InventoryAction(NetworkPacket* pkt)
...
-
-		bool to_inv_is_current_player =
-			(ma->to_inv.type == InventoryLocation::PLAYER) &&
-			(ma->to_inv.name == player->getName());
+		bool from_inv_is_current_player = false;
+		if (ma->from_inv.type == InventoryLocation::PLAYER) {
+			if (ma->from_inv.name != player->getName())
+				return;
+			from_inv_is_current_player = true;
+		} else if (ma->from_inv.type != InventoryLocation::CURRENT_PLAYER)
```
Comment (sfan5) [inline src/network/serverpackethandler.cpp:637]: should be `ma->to_inv.type != InventoryLocation::CURRENT_PLAYER`
you can also move that to the same line as the declaration (`bool foo = x == y;`)

File: src/network/serverpackethandler.cpp [hunk 3]
```diff
@@ -629,13 +629,19 @@ void Server::handleCommand_InventoryAction(NetworkPacket* pkt)
...
 			m_inventory_mgr->setInventoryModified(ma->to_inv);
 
-		bool from_inv_is_current_player =
-			(ma->from_inv.type == InventoryLocation::PLAYER) &&
-			(ma->from_inv.name == player->getName());
-
-		bool to_inv_is_current_player =
-			(ma->to_inv.type == InventoryLocation::PLAYER) &&
-			(ma->to_inv.name == player->getName());
+		bool from_inv_is_current_player = ma->from_inv.type != InventoryLocation::CURRENT_PLAYER;
```
Comment (sfan5) [inline src/network/serverpackethandler.cpp:632]: This is still incorrect
the action is from current players inventory if `ma->from_inv.type == InventoryLocation::CURRENT_PLAYER`

Comment (appgurueu) [reply src/network/serverpackethandler.cpp]: Sorry, should be fixed now

**Review by sfan5 — APPROVED** [review]

File: src/network/serverpackethandler.cpp [hunk 4]
```diff
@@ -629,13 +629,19 @@ void Server::handleCommand_InventoryAction(NetworkPacket* pkt)
...
 			m_inventory_mgr->setInventoryModified(ma->to_inv);
 
-		bool from_inv_is_current_player =
-			(ma->from_inv.type == InventoryLocation::PLAYER) &&
-			(ma->from_inv.name == player->getName());
-
-		bool to_inv_is_current_player =
-			(ma->to_inv.type == InventoryLocation::PLAYER) &&
-			(ma->to_inv.name == player->getName());
+		bool from_inv_is_current_player = ma->from_inv.type == InventoryLocation::CURRENT_PLAYER;
```
Comment (SmallJoker) [inline src/network/serverpackethandler.cpp:632]: `CURRENT_PLAYER` is only used for client-side applied changes and is ignored server-side. See serveinventorymgr.cpp L44 for example. Though it does not break anything, this check is superfluous.

Comment (sfan5) [reply src/network/serverpackethandler.cpp:632]: I see, I should've checked this. The line can be changed to `= false;` then

**Comment by rubenwardy** [comment]
My approval still stands

**Comment by anon55555** [comment]
Nice, the inventory exploit I found is patched.

**Comment by rubenwardy** [comment]
We'd be grateful for reports in the future, I know that's less fun :)

**Comment by anon55555** [comment]
I emailed sfan5

> We'd be grateful for reports in the future, I know that's less fun :D

I emailed a core dev about it instead of reporting it publicly because I didn't want somebody wiping everyone's inventories on servers.

**Comment by anon55555** [comment]
> We'd be grateful for reports in the future, I know that's less fun :D

I emailed a core dev about it instead of reporting it publicly because I didn't want somebody seeing the issue and wiping everyone's inventories on servers. But it seems the email wasn't noticed. Is there a way to report bugs privately that gets noticed?

**Comment by sfan5** [comment]
Talking to coredevs privately on IRC is another option, whne you get a reply you can be sure your report was seen.

**Comment by rubenwardy** [comment]
Ah ok, thanks for that - we need to work on our communication

**Comment by LoneWolfHT** [comment]
https://github.com/minetest/minetest/security/policy

Thought that tab allowed in-Github reports, guess not

**Comment by rubenwardy** [comment]
Oh nice, you can draft the advisory privately and then publish it

**Comment by Lejo1** [comment]
Needs to be done for detached inventorys as well.
3d_armor has a workaround but they should be safe by the engine.

**Comment by appgurueu** [comment]
> Needs to be done for detached inventorys as well.
> 3d_armor has a workaround but they should be safe by the engine.

Impossible. Detached inventories are meant to be accessible. Only mods can and have to implement meaningful restrictions, such as range or ownership.

**Comment by Lejo1** [comment]
I'm referring to detached inventorys which were specifically meant for only one player. These are also only sent to this one.
Sure public detached inventorys should still be accessible.

**Comment by SmallJoker** [comment]
#11035

**Pull request merged**